### PR TITLE
[🍍 < T1157-MG] Support concurrent RsmClient requests

### PR DESCRIPTION
**Pull request by spacejam**
Add support for multiple concurrent requests in the RsmClient, accessible via a new `AsyncRequestToken`. Update the ShardRequestManager to use the new request tokens and refactor some bug-prone aspects of the existing logic.

[master < Epic] PR
- [ ] Check, and update documentation if necessary
- [ ] Update [changelog](https://docs.memgraph.com/memgraph/changelog)
- [ ] Write E2E tests
- [ ] Compare the [benchmarking results](https://bench-graph.memgraph.com/) between the master branch and the Epic branch
- [ ] Provide the full content or a guide for the final git message

[master < Task] PR
- [ ] Check, and update documentation if necessary
- [ ] Update [changelog](https://docs.memgraph.com/memgraph/changelog)
- [ ] Provide the full content or a guide for the final git message

### CODE REVIEW

**Review by 42jeremy — APPROVED** [review]

**Review by 42jeremy — APPROVED** [review]

File: tests/unit/high_density_shard_create_scan.cpp [hunk 1]
```diff
@@ -196,7 +194,7 @@ void ExecuteOp(query::v2::RequestRouter<LocalTransport> &request_router, std::se
                ScanAll scan_all) {
   query::v2::ExecutionState<msgs::ScanVerticesRequest> request{.label = "test_label"};
```
Comment (kostasrim) [inline tests/unit/high_density_shard_create_scan.cpp:195]: dead code

File: src/query/v2/request_router.hpp [hunk 2]
```diff
@@ -246,99 +237,116 @@ class RequestRouter : public RequestRouterInterface {
...
-    } while (!all_requests_gathered(paginated_response_tracker));
+      DriveReadResponses(state, responses);
+    } while (!state.requests.empty());
 
-    MaybeCompleteState(state);
-    // TODO(kostasrim) Before returning start prefetching the batch (this shall be done once we get MgFuture as return
-    // result of storage_client.SendReadRequest()).
-    return PostProcess(std::move(responses));
+    // convert responses into VertexAccessor objects to return
+    std::vector<VertexAccessor> accessors;
```
Comment (kostasrim) [inline src/query/v2/request_router.hpp:262]: reserve the size ?

File: src/io/rsm/rsm_client.hpp [hunk 3]
```diff
@@ -91,162 +101,145 @@ class RsmClient {
...
+
+    WriteRequest<WriteRequestT> write_req = {.operation = async_request.request};
+
+    async_request.future =
+        io_.template Request<WriteRequest<WriteRequestT>, WriteResponse<WriteResponseT>>(leader_, write_req);
   }
 
-  std::optional<BasicResult<TimedOut, WriteResponseT>> PollAsyncWriteRequest() {
-    MG_ASSERT(async_write_);
+  std::optional<BasicResult<TimedOut, WriteResponseT>> PollAsyncWriteRequest(AsyncRequestToken &token) {
```
Comment (kostasrim) [inline src/io/rsm/rsm_client.hpp:218]: `AsyncRequestToken &token` can be made `const`

File: src/io/rsm/rsm_client.hpp [hunk 4]
```diff
@@ -91,162 +101,145 @@ class RsmClient {
...
       return std::nullopt;
     }
 
     return AwaitAsyncWriteRequest();
   }
 
-  std::optional<BasicResult<TimedOut, WriteResponseT>> AwaitAsyncWriteRequest() {
-    ResponseResult<WriteResponse<WriteResponseT>> get_response_result = std::move(*async_write_).Wait();
-    async_write_.reset();
+  std::optional<BasicResult<TimedOut, WriteResponseT>> AwaitAsyncWriteRequest(AsyncRequestToken &token) {
```
Comment (kostasrim) [inline src/io/rsm/rsm_client.hpp:228]: Same, `const` here

**Review by kostasrim — APPROVED** [review]

File: tests/unit/high_density_shard_create_scan.cpp (hunk 1 above)
Comment (spacejam) [reply tests/unit/high_density_shard_create_scan.cpp]: removed

File: src/io/rsm/rsm_client.hpp (hunk 3 above)
Comment (spacejam) [reply src/io/rsm/rsm_client.hpp]: applied for all methods that accept a token argument

File: src/io/rsm/rsm_client.hpp (hunk 4 above)
Comment (spacejam) [reply src/io/rsm/rsm_client.hpp]: applied, thanks

**Pull request merged**